File: 14.1.clj
(def w 101)
(def h 103)
(def qw (quot w 2))
(def qh (quot h 2))

(def robots(->>
    (slurp (first *command-line-args*))
    (re-seq #"p=(\d+),(\d+) v=(-*\d+),(-*\d+)")
    (map #(subvec % 1 5))
    (map #(vec (map Integer/parseInt %)))
    (map #(
        vector
            (mod (+ (% 0) (* 100 (% 2))) w)
            (mod (+ (% 1) (* 100 (% 3))) h)
    ))
))

(def qlu (count (filter (fn [[x y]] (and (< x qw) (< y qh) )) robots)))
(def qru (count (filter (fn [[x y]] (and (> x qw) (< y qh) )) robots)))
(def qld (count (filter (fn [[x y]] (and (< x qw) (> y qh) )) robots)))
(def qrd (count (filter (fn [[x y]] (and (> x qw) (> y qh) )) robots)))

(println (= (* qlu qru qld qrd) 231782040))

File: 14.1.py
import sys
import re

w = 101
h = 103
qw = w // 2
qh = h // 2

input = open(sys.argv[1], "r+t").read()
robots = re.findall(r"p=(\d+),(\d+) v=(-*\d+),(-*\d+)", input)
robots = [ list(map(int, r)) for r in robots ]
robots_after_100_seconds = [
    [
        (x + 100*wx) % w,
        (y + 100*wy) % h
    ]
    for x, y, wx, wy in robots
]

qlu = len([1 for x, y in robots_after_100_seconds if x < qw and y < qh ])
qru = len([1 for x, y in robots_after_100_seconds if x > qw and y < qh ])
qld = len([1 for x, y in robots_after_100_seconds if x < qw and y > qh ])
qrd = len([1 for x, y in robots_after_100_seconds if x > qw and y > qh ])

print(qlu*qru*qld*qrd == 231782040)
